### PR TITLE
feat(timeselect): Add inline dp-time-select

### PR DESCRIPTION
**Pull request by isaacplmann**
I've added an inline `<dp-time-select>` component.  Here's the config object:

```ts
export interface ITimeSelectConfig extends ICalendar {
  hours12Format?: string; // 'hh'
  hours24Format?: string; // 'HH'
  meridiemFormat?: string; // 'A'
  minutesFormat?: string; // 'mm'
  minutesInterval?: number; // 1
  secondsFormat?: string; // 'ss'
  secondsInterval?: number; // 1
  showSeconds?: boolean; // false
  showTwentyFourHours?: boolean; // false
  timeSeparator?: string; // ':'
}
```

Next step is to add values to the ECalendarType enum: `Time` and `DayTime` (or `DateTime`) and then add the appropriate integrations in the `<dp-day-picker>`.  I'll work on `Time` first, since it is the simpler one.

Discussion points:
1. I can foresee some ambiguity with the `min` and `max` values in the config object.  For a `Time` calendar type, the `min` and `max` should probably have the day portion of the datetime ignored and only compare the time portion with the value in the `<dp-time-select>`.  However, for a `DayTime` calendar type, the day portion needs to be included in the calculation.  I think it is clear enough from the context what `min` and `max` should mean, but, if you want, I can add `minTime` and `maxTime` properties to the `ITimeSelectConfig`.

2. It would probably be useful to update the demo code to hide irrelevant config options based on the currently selected mode.  I was thinking of creating some kind of look up table to hide/show config options as the mode changes.  I'd be happy to work on that if you think it's a good idea.

### CODE REVIEW

**Comment by isaacplmann** [comment]
Starts to implement #5

**Comment by vlio20** [comment]
Hi @isaacplmann,
1. I think it is a good idea to support `minTime/maxTime` if user don't want to allow specific hours for a day.
2. sure - sounds good.

Few questions and points that have to be added: 
1. Disabled options.
2. there should be an option to have validation errors for min and max time (not only disabled state).
3. Material theme also should be supported.
4. Can it be placed next to day picker? - I think it should (how will it look?).
5. what about a directive (sees like many of the users a using this option (according to open issues), will you also add it?

**Comment by isaacplmann** [comment]
3. What else should I do for the material theme?  I change the buttons to being circles that appear on hover.
4. Yes, I'm planning on implementing that.
5. Yes, I'll add a directive version.

**Comment by isaacplmann** [comment]
Todo list:

- [ ] `minTime`/`maxTime`
- [ ] show only relevant config options in demo
- [ ] disabled options
- [ ] validation errors for `minTime`/`maxTime`
- [ ] `Time` version of `dp-day-picker` (opens a `time-select`)
- [ ] inline `DayTime` - (`time-select` and `day-calendar` side by side)
- [ ] `DayTime` version of `dp-day-picker`
- [ ] directive version of `Time` calendar type
- [ ] directive version of `DayTime` calendar type

**Comment by vlio20** [comment]
3. Can you please share a screenshot?

**Comment by isaacplmann** [comment]
Material theme (hovering over the bottom middle button)
screen shot 2017-06-26 at 9 20 08 am

No theme
screen shot 2017-06-26 at 9 20 20 am

**Comment by isaacplmann** [comment]
Travis CI is failing, but I think that's happening in code that I haven't touched.  Any thoughts?  When I run `npm test` all the tests pass.

**Comment by vlio20** [comment]
will check

**Comment by vlio20** [comment]
@isaacplmann, can you also deploy the demo to your forked project?

File: src/app/common/services/utils/utils.service.ts [hunk 1]
```diff
@@ -6,6 +6,14 @@ import {Moment, unitOfTime} from 'moment';
 import {CalendarValue} from '../../types/calendar-value';
 import {IDate} from '../../models/date.model';
 
+export type DateValidatorFn = (inputVal: CalendarValue) => {[key: string]: any};
```
Comment (vlio20) [inline src/app/common/services/utils/utils.service.ts]: why any?

File: src/app/demo/demo/demo.component.html [hunk 2]
```diff
@@ -127,6 +128,38 @@
...
       </div>
 
+      <div class="dp-inline" *ngIf="pickerMode === 'timeInline'">
+        <dp-time-select
+            name="timeInline"
+            #timeInline="ngModel"
+            [theme]="material ? 'dp-material' : ''"
+            [(ngModel)]="date"
+            (ngModelChange)="log($event)"
+            [minDate]="validationMinDate"
```
Comment (vlio20) [inline src/app/demo/demo/demo.component.html]: why the time component has min/maxDate validations?

File: src/app/time-select/time-select.component.html [hunk 3]
```diff
@@ -0,0 +1,28 @@
+<div class="dp-time-select-container">
+  <ul class="controls">
+    <li class="control">
+      <button [disabled]="!shouldShowIncrease('hour')" class="control-up" (click)="increase('hour')"></button>
```
Comment (vlio20) [inline src/app/time-select/time-select.component.html]: all classes should be prefixed with dp

File: src/app/time-select/time-select.component.html [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      <span class="control--hours">{{hours}}</span>
```
Comment (vlio20) [inline src/app/time-select/time-select.component.html]: I am not using BEM, so please remove

File: src/app/time-select/time-select.component.less [hunk 5]
```diff
@@ -0,0 +1,55 @@
+@import '../common/styles/variables';
+
+& {
+  .controls {
```
Comment (vlio20) [inline src/app/time-select/time-select.component.less]: please add line break after each class

File: src/app/time-select/time-select.component.less [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    letter-spacing: 1.3px;
+  }
+  .control-up, .control-down {
+    position: relative;
+    display: block;
+    margin: 3px auto;
+    font-size: 18px;
+    font-weight: bold;
+    cursor: pointer;
+    &:before {
```
Comment (vlio20) [inline src/app/time-select/time-select.component.less]: : =-> ::

File: src/app/time-select/time-select.component.less [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+      display: inline-block;
+      position: relative;
+    }
+  }
+  .control-up:before {
+    content: '︿';
```
Comment (vlio20) [inline src/app/time-select/time-select.component.less]: We shouldn't use this type of characters as they are probably not supported on linux or other OSs. Please use pure CSS.

File: src/app/time-select/time-select.component.less [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    bottom: 0.2em;
```
Comment (vlio20) [inline src/app/time-select/time-select.component.less]: why em? please use px

File: src/app/time-select/time-select.component.ts [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    this.selected = this.selected || moment();
+    this.inputValueType = this.utilsService.getInputType(this.inputValue, false);
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (this.isInited) {
+      const {minDate, maxDate, minTime, maxTime} = changes;
+      this.init();
+
+      if (minDate || maxDate || minTime || maxTime) {
```
Comment (vlio20) [inline src/app/time-select/time-select.component.ts]: please push this check into the `initValidatiors` method

File: src/app/time-select/time-select.component.ts [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+      useExisting: forwardRef(() => TimeSelectComponent),
+      multi: true
+    }
+  ]
+})
+export class TimeSelectComponent implements OnInit, OnChanges, ControlValueAccessor, Validator {
+
+  @Input() config: ITimeSelectConfig;
+  @Input() displayDate: SingleCalendarValue;
+  @Input() minDate: Moment;
```
Comment (vlio20) [inline src/app/time-select/time-select.component.ts]: we should also support strings, that means that this component should also receive a format as an input.

File: src/app/time-select/time-select.service.ts [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    maxTime: undefined,
+    min: undefined,
+    minTime: undefined,
+    minutesFormat: 'mm',
+    minutesInterval: 1,
+    secondsFormat: 'ss',
+    secondsInterval: 1,
+    showSeconds: false,
+    showTwentyFourHours: false,
+    timeSeparator: ':',
```
Comment (vlio20) [inline src/app/time-select/time-select.service.ts]: redundant ,

File: src/app/time-select/time-select.service.ts [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+  constructor(private utilsService: UtilsService) {
+  }
+
+  getConfig(config: ITimeSelectConfig): ITimeSelectConfig {
+    const timeConfigs = {
+      maxTime: this.utilsService.onlyTime(config && config.maxTime),
+      minTime: this.utilsService.onlyTime(config && config.minTime),
+    };
+    return {...this.DEFAULT_CONFIG, ...this.utilsService.clearUndefined(config), ...timeConfigs};
```
Comment (vlio20) [inline src/app/time-select/time-select.service.ts]: please break to lines

File: src/app/time-select/time-select.service.ts [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    timeSeparator: ':',
+  };
+
+  constructor(private utilsService: UtilsService) {
+  }
+
+  getConfig(config: ITimeSelectConfig): ITimeSelectConfig {
+    const timeConfigs = {
+      maxTime: this.utilsService.onlyTime(config && config.maxTime),
+      minTime: this.utilsService.onlyTime(config && config.minTime),
```
Comment (vlio20) [inline src/app/time-select/time-select.service.ts]: redundant ,
I will add lint rule for that in the near future

**Comment by isaacplmann** [comment]
How do you deploy the demo?  I see a script to build the demo, but not to deploy it.

**Comment by vlio20** [comment]
1. go to release branch (dist folder is not ignored there) 
2. `run npm run build:demo`
3. run: `git subtree push --prefix dist origin gh-pages`

**Comment by vlio20** [comment]
@isaacplmann, I pushed a fix to master for the travis build there is still an issue with one e2e tests which is not related to your code.
I will try to fix it later.

File: src/app/demo/demo/demo.component.html (hunk 2 above)
Comment (isaacplmann) [inline src/app/demo/demo/demo.component.html]: You're right, I should put the `minDate` and `maxDate` validation only on the `DayTime` component and not on the `Time` component.

File: src/app/time-select/time-select.component.ts (hunk 9 above)
Comment (isaacplmann) [inline src/app/time-select/time-select.component.ts]: That check doesn't make sense in the `initValidators` method.  It's checking to see if those properties have been changed - which only makes sense in `ngOnChanges`. See https://github.com/vlio20/angular-datepicker/blob/master/src/app/day-calendar/day-calendar.component.ts#L119

File: src/app/time-select/time-select.service.ts (hunk 11 above)
Comment (isaacplmann) [inline src/app/time-select/time-select.service.ts]: I'm modeling the `DEFAULT_CONFIG` style off of `day-calendar.service` and `month-calendar.service`.  What's redundant?

Comment (isaacplmann) [inline src/app/time-select/time-select.service.ts]: Ah, you're referring to the comma.  Ok, I'll remove it.

File: src/app/time-select/time-select.component.ts (hunk 9 above)
Comment (vlio20) [inline src/app/time-select/time-select.component.ts]: you are totally right - my mistake

**Comment by isaacplmann** [comment]
Demo is up: https://isaacplmann.github.io/ng2-date-picker/

**Comment by isaacplmann** [comment]
Made the buttons css only - and I think they look better too.

File: src/app/time-select/time-select.service.ts (hunk 11 above)
Comment (vlio20) [inline src/app/time-select/time-select.service.ts]: yup

**Comment by isaacplmann** [comment]
Created `daytimeInline` and `daytimePicker` modes.  Next I'll work on `timeDirective` and `daytimeDirective`.

**Comment by isaacplmann** [comment]
Directive versions are finished.  I've updated the demo too.

**Comment by isaacplmann** [comment]
You can check out the new docs here: https://github.com/isaacplmann/ng2-date-picker/tree/timepicker

I'm done with my changes - I'm just waiting on your review.

**Comment by vlio20** [comment]
can you please add the demo again - seems that it is not working any more ...

**Comment by isaacplmann** [comment]
Here's a hosted version of the demo: https://isaacplmann.github.io/ng2-date-picker/
I also just ran `npm start` on my local copy and it served up the demo.  What isn't working for you?

File: src/app/time-select/time-select.service.ts [hunk 14]
```diff
@@ -0,0 +1,128 @@
+import {Injectable} from '@angular/core';
+import {Moment} from 'moment';
+import * as moment from 'moment';
+
```
Comment (vlio20) [inline src/app/time-select/time-select.service.ts]: redundant

**Comment by vlio20** [comment]
@isaacplmann, the demo looks great!!! 
one thing though is with the no theme appearance - the button icon is not in the middle: 
Imgur

Also I would prefer to make it more slim: 
Imgur

**Comment by vlio20** [comment]
another question: what is `package-lock.json`?  I was using shrinkwrap.

**Comment by isaacplmann** [comment]
`package-lock.json` is a file generated by `npm@5.x`.  It's the next version of shrinkwrap.  I'll remove it from the PR, I must have accidentally included it.

**Comment by isaacplmann** [comment]
I'll fix the e2e tests and add some for the new components tomorrow.

**Comment by isaacplmann** [comment]
Fixed e2e tests and added some for `timePicker` and `daytimePicker`.

**Comment by vlio20** [comment]
This is awesome! Can you just make sure that the default state in the demo is the date picker with the time selector?

**Comment by isaacplmann** [comment]
You want the default `pickerMode` to be `daytimePicker`?

**Comment by isaacplmann** [comment]
If that's the case, then the default Date Format should be set to `DD-MM-YYYY HH:mm:ss`

**Comment by vlio20** [comment]
Yes, it's shows better the capabilities of this component. 
There probably should be default format for each type of calendar - so for there date time picker it should be `Yes `DD-MM-YYYY HH:mm:ss` as you mentioned.

**Comment by isaacplmann** [comment]
I updated `pickerMode` to use `daytimePicker` by default and fixed the e2e tests that were broken by that change.  I'll leave modifying the `config.format` when the `pickerMode` changes to you.

**Comment by vlio20** [comment]
cool, so you are done?

**Comment by isaacplmann** [comment]
Yep.  All done.

**Comment by vlio20** [comment]
I will release a version soon. thanks again.

**Comment by isaacplmann** [comment]
Thank you!

**Pull request merged**